src/adapter/src/catalog/apply.rs

```diff
@@ -1092,6 +1024,113 @@ impl CatalogState {
     }
 }
 
+/// Most updates are applied one at a time, but during bootstrap, certain types are applied
+/// separately in a batch for performance reasons. A constraint is that updates must be applied in
+/// order. This method is modeled as a state machine that batches then applies groups of updates.
+enum BootstrapApplyState {
+    /// Additions of builtin views.
+    BuiltinViewAdditions(Vec<(&'static Builtin<NameReference>, GlobalId)>),
+    /// Item updates that aren't builtin view additions.
+    Items(Vec<StateUpdate>),
+    /// All other updates.
+    Updates(Vec<StateUpdate>),
+}
+
+impl BootstrapApplyState {
+    fn new(update: StateUpdate) -> BootstrapApplyState {
+        match update {
+            StateUpdate {
+                kind: StateUpdateKind::SystemObjectMapping(system_object_mapping),
+                diff: StateDiff::Addition,
+            } if matches!(
+                system_object_mapping.description.object_type,
+                CatalogItemType::View
+            ) =>
+            {
+                let view_addition = lookup_builtin_view_addition(system_object_mapping);
+                BootstrapApplyState::BuiltinViewAdditions(vec![view_addition])
+            }
+            StateUpdate {
+                kind: StateUpdateKind::IntrospectionSourceIndex(_),
+                ..
+            }
+            | StateUpdate {
+                kind: StateUpdateKind::SystemObjectMapping(_),
+                ..
+            }
+            | StateUpdate {
+                kind: StateUpdateKind::Item(_),
+                ..
+            } => BootstrapApplyState::Items(vec![update]),
+            update => BootstrapApplyState::Updates(vec![update]),
+        }
+    }
+
+    async fn apply(
+        &mut self,
+        state: &mut CatalogState,
+        retractions: &mut InProgressRetractions,
+    ) -> Vec<BuiltinTableUpdate<&'static BuiltinTable>> {
+        match self {
+            BootstrapApplyState::BuiltinViewAdditions(builtin_view_additions) => {
+                // We enable all "enable_for_item_parsing" feature flags when applying item updates
+                // during bootstrap.
+                //
+                // See [`CatalogState::with_enable_for_item_parsing`] for more details.
+                let restore = state.system_configuration.clone();
+                state.system_configuration.enable_for_item_parsing();
+                let builtin_table_updates =
+                    Catalog::parse_builtin_views(state, std::mem::take(builtin_view_additions))
+                        .await;
+                state.system_configuration = restore;
+                builtin_table_updates
```

Comment:
I'm on the fence about whether the `system_configuration` stuff belongs here or in `Catalog::parse_builtin_views`. Let me know your opinions.

Reply:
I don't have an opinion here. If working on the code didn't produce a clear answer, perhaps it doesn't matter much.

Reply:
I'll leave as is then.